src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/util/AzureStorageHelper.java

```diff
@@ -0,0 +1,125 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(AzureStorageHelper.class);
+
+    @Autowired
+    public AzureStorageHelper(ServiceURL client) {
+        this.client = client;
+    }
+
+    public ServiceURL getClient() {
+        return client;
+    }
+
+    public Single<ServiceListContainersSegmentResponse> listContainers() {
+        return client.listContainersSegment(null, ListContainersOptions.DEFAULT, null)
+            .flatMap(this::listContainers);
+    }
+
+    public Observable<BlobItem> listBlobsLazy(ContainerURL containerURL) {
+        return containerURL.listBlobsFlatSegment(null, null, null)
+            .flatMapObservable((r) -> listContainersResultToContainerObservable(containerURL, null, r));
+    }
+
+    private static Observable<BlobItem> listContainersResultToContainerObservable(
+        ContainerURL containerURL, ListBlobsOptions listBlobsOptions,
+        ContainerListBlobFlatSegmentResponse response
+    ) {
+        BlobFlatListSegment segment = response.body().segment();
+        if (segment != null) {
+            Observable<BlobItem> result = Observable.fromIterable(segment.blobItems());
+
+            LOGGER.info("Count: {}", segment.blobItems());
+
+            if (response.body().nextMarker() != null) {
+                LOGGER.info("Hit continuation in listing at {}", segment.blobItems().get(
+                    segment.blobItems().size() - 1).name());
+                // Recursively add the continuation items to the observable.
+                result = result
+                    .concatWith(containerURL.listBlobsFlatSegment(response.body().nextMarker(), listBlobsOptions, null)
+                        .flatMapObservable((r) ->
+                            listContainersResultToContainerObservable(containerURL, listBlobsOptions, r)));
+            }
+
+            return result;
+        }
+
+        return Observable.empty();
+    }
+
+    private Single<ServiceListContainersSegmentResponse> listContainers(
+        ServiceListContainersSegmentResponse response
+    ) {
+        // If there is not another segment, return this response as the final response.
+        if (response.body().nextMarker() == null) {
+            return Single.just(response);
+        } else {
+            /*
+             IMPORTANT: ListContainersSegment returns the start of the next segment; you MUST use this to get the
+             next segment (after processing the current result segment
+             */
+            String nextMarker = response.body().nextMarker();
+
+            /*
+            The presence of the marker indicates that there are more blobs to list, so we make another call to
+            listContainersSegment and pass the result through this helper function.
+             */
+            return client.listContainersSegment(nextMarker, ListContainersOptions.DEFAULT, null)
+                .flatMap(containersListBlobHierarchySegmentResponse -> listContainers(response));
```

Comment:
is the list containers supposex to use `containersListBlobHierarchySegmentResponse` or `response` ?

Reply:
not really sure, thats from the ms sample code here:
https://github.com/Azure/azure-storage-java/blob/master/src/test/java/com/microsoft/azure/storage/Samples.java